src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs

```diff
@@ -0,0 +1,846 @@
+            {
+                Debug.Assert(chosenItems.Length > 0);
+
+                // Try to find the chosen item has been most recently used.
+                var bestItem = chosenItems[0];
+                for (int i = 1, n = chosenItems.Length; i < n; i++)
+                {
+                    var chosenItem = chosenItems[i];
+                    var mruIndex1 = GetRecentItemIndex(recentItems, bestItem);
+                    var mruIndex2 = GetRecentItemIndex(recentItems, chosenItem);
+
+                    if ((mruIndex2 < mruIndex1) ||
+                        (mruIndex2 == mruIndex1 && !bestItem.IsPreferredItem() && chosenItem.IsPreferredItem()))
+                    {
+                        bestItem = chosenItem;
+                    }
+                }
+
+                // If our best item appeared in the MRU list, use it
+                if (GetRecentItemIndex(recentItems, bestItem) <= 0)
+                {
+                    return bestItem;
+                }
+
+                // Otherwise use the chosen item that has the highest
+                // matchPriority.
+                for (int i = 1, n = chosenItems.Length; i < n; i++)
+                {
+                    var chosenItem = chosenItems[i];
+                    var bestItemPriority = bestItem.Rules.MatchPriority;
+                    var currentItemPriority = chosenItem.Rules.MatchPriority;
+
+                    if ((currentItemPriority > bestItemPriority) ||
+                        ((currentItemPriority == bestItemPriority) && !bestItem.IsPreferredItem() && chosenItem.IsPreferredItem()))
+                    {
+                        bestItem = chosenItem;
+                    }
+                }
+
+                return bestItem;
+
+                static int GetRecentItemIndex(ImmutableArray<string> recentItems, RoslynCompletionItem item)
+                {
+                    var index = recentItems.IndexOf(item.FilterText);
+                    return -index;
+                }
+            }
+
+            private static bool TryGetInitialTriggerLocation(AsyncCompletionSessionDataSnapshot data, out SnapshotPoint intialTriggerLocation)
+            {
+                var firstItem = data.InitialSortedList.FirstOrDefault(static item => item.Properties.ContainsProperty(CompletionSource.TriggerLocation));
+                if (firstItem != null)
+                {
+                    return firstItem.Properties.TryGetProperty(CompletionSource.TriggerLocation, out intialTriggerLocation);
+                }
+
+                intialTriggerLocation = default;
+                return false;
+            }
+
+            private static bool IsHardSelection(
+                string filterText,
+                RoslynCompletionItem item,
+                bool matchedFilterText,
+                bool useSuggestionMode)
+            {
+                if (item == null || useSuggestionMode)
```

Comment:
odd null check given that this file is NRT enabled, and `item` is not marked as being nullable.